### ISSUE TITLE
PIO issues with MPICH 4.x

### ISSUE THREAD

**Issue by theurich** [esmf-org/esmf#107]
We do not currently test with MPICH > 3. From a user report under ticket https://github.com/esmf-org/esmf-support/issues/157 (see email traffic!!!) we expect that there will be issues. Specifically the user had to apply a patch to PIO code - which the user supplied (see email!). 

Keep this in mind when we start testing with MPICH 4.x and above.

**Comment by theurich** [comment]
I added testing under my laptop (breve) for gfortran 11.2.0 + MPICH 4.0.0. There are indeed a number of failures with this combo as shown under [summary.dat](https://github.com/esmf-org/esmf-test-artifacts/blob/f5b77ae30e344371f308a3bd65c553aeac38e6a0/develop/gfortran/11.2.0/g/mpich/4.0.0/summary.dat). 

Test [test.log](https://raw.githubusercontent.com/esmf-org/esmf-test-artifacts/f5b77ae30e344371f308a3bd65c553aeac38e6a0/develop/gfortran/11.2.0/g/mpich/4.0.0/out/test.log) shows:

```
The following unit test files failed to build, failed to execute or crashed during execution:

CRASHED: mpich/g: src/Infrastructure/Array/tests/ESMF_ArrayIOUTest.F90
CRASHED: mpich/g: src/Infrastructure/ArrayBundle/tests/ESMF_ArrayBundleIOUTest.F90
CRASHED: mpich/g: src/Infrastructure/Field/tests/ESMC_FieldSMMFromFileUTest.c
CRASHED: mpich/g: src/Infrastructure/Field/tests/ESMF_FieldIOUTest.F90
CRASHED: mpich/g: src/Infrastructure/Field/tests/ESMF_FieldSMMFromFileUTest.F90
CRASHED: mpich/g: src/Infrastructure/FieldBundle/tests/ESMF_FieldBundleIOUTest.F90
CRASHED: mpich/g: src/Infrastructure/IO/tests/ESMF_IOUTest.F90
CRASHED: mpich/g: src/Infrastructure/IO/tests/ESMF_IO_MultitileUTest.F90
CRASHED: mpich/g: src/Infrastructure/Mesh/tests/ESMF_MeshFileIOUTest.F90
CRASHED: mpich/g: src/Infrastructure/Route/tests/ESMF_RouteHandleAdvancedUTest.F90
CRASHED: mpich/g: src/Superstructure/PreESMFMod/tests/ESMF_FileRegridUTest.F90
CRASHED: mpich/g: src/Superstructure/PreESMFMod/tests/ESMF_RegridWeightGenUTest.F90
CRASHED: mpich/g: src/addon/NUOPC/tests/ESMF_NUOPC_UTest.F90
```
and
```
The following examples failed, did not build, or did not execute:

FAIL: mpich/g: src/Infrastructure/Mesh/examples/ESMF_MeshEx.F90
FAIL: mpich/g: src/Infrastructure/Route/examples/ESMF_RHandleBitForBitEx.F90
```

All these tests are I/O related, which seems to align with the report from the user mentioned previously where there was an issue with PIO and MPICH 4.x reported.

Drilling into the ESMF_IOUTest failure:

- [ESMF_IOUTest.stdout](https://github.com/esmf-org/esmf-test-artifacts/blob/f5b77ae30e344371f308a3bd65c553aeac38e6a0/develop/gfortran/11.2.0/g/mpich/4.0.0/test/ESMF_IOUTest.stdout)
- [ESMF_IOUTest.Log](https://github.com/esmf-org/esmf-test-artifacts/blob/f5b77ae30e344371f308a3bd65c553aeac38e6a0/develop/gfortran/11.2.0/g/mpich/4.0.0/test/ESMF_IOUTest.Log)

**Comment by theurich** [comment]
This looks like an I/O issue, therefore assigned to @jedwards4b and @billsacks. In particular it looks down in PIO. It will be interesting to find out if PIO has been tested with MPICH 4.x yet.

**Comment by theurich** [comment]
Changed the title to reflect the current status of this.

**Comment by jedwards4b** [comment]
I think that this may be a reoccurring problem with mpi implementations rearing its ugly head again.  MPI_Alltoallw should allow the data type to be MPI_DATATYPE_NULL.    I will test and possibly send a bug report to mpich developers.

**Comment by theurich** [comment]
Thanks @jedwards4b for that thread under OpenMPI. Very interesting to scan through. So in the end I guess they changed the behavior of the implementation, even after having pointed out that the standard does not require them to accept MPI_DATATYPE_NULL for count of 0?

**Comment by jedwards4b** [comment]
I think that's what happened but after reading the full text I'm not completely sure.  I'll need to look into it further.

**Comment by jedwards4b** [comment]
I found this in pio_internal.h - this is apparently how I dealt with the issue 7 years ago.  
/** Some MPI implementations do not allow passing MPI_DATATYPE_NULL to                                                                                                                                   
 * comm functions even though the send or recv length is 0, in these                                                                                                                                     
 * cases we use MPI_CHAR */                                                                                                                                                                              
#if defined(MPT_VERSION) || defined(OPEN_MPI)                                                                                                                                                            
#define PIO_DATATYPE_NULL MPI_CHAR                                                                                                                                                                       
#else                                                                                                                                                                                                    
#define PIO_DATATYPE_NULL MPI_DATATYPE_NULL                                                                                                                                                              
#endif                          

Can you try again after commenting the code so that
#define PIO_DATATYPE_NULL MPI_CHAR       
is always used.

**Comment by theurich** [comment]
@jedwards4b, I am re-building, and re-running the full test suite now. Will let you know.

**Comment by jedwards4b** [comment]
https://github.com/pmodels/mpich/issues/6323

**Comment by theurich** [comment]
@jedwards4b, that fixed the issue. [The results are clean now](https://github.com/esmf-org/esmf-test-artifacts/blob/f3fe8789d81babd382cb95f6d2362d5a58bf2d7d/fix_pio-alltoallw/gfortran/11.2.0/O/mpich/4.0.0/summary.dat).

**Comment by theurich** [comment]
@jedwards4b, do you see any downside to the change under https://github.com/esmf-org/esmf/commit/46b3caf9f031a08c3032ab29eb3c4c7cc253351e?

**Comment by jedwards4b** [comment]
No, I think that that's what I'll implement for the next PIO tag.  I think that MPI libraries really ought to support 
this usage of MPI_DATATYPE_NULL but I guess I had better just give up on the idea.

**Comment by theurich** [comment]
The issue has officially been resolved with PIO 2.5.10. This version of PIO has been merged into ESMF with https://github.com/esmf-org/esmf/pull/111.
